chain/store/name.go

```diff
@@ -1,63 +1,177 @@
+}
 
-	return sdb.trie.TryUpdate(append(NameRegistrantPrefix, registrantId...), []byte(name))
+func (ni *nameInfo) String() string {
+	return fmt.Sprintf("%064x %d", ni.registrant, ni.expiresAt)
 }
 
-func (sdb *StateDB) setName(registrant []byte, name string) {
-	registrantId := getRegistrantId(registrant)
-	nameId := getNameId(name)
+func (ni *nameInfo) Empty() bool {
+	return len(ni.registrant) == 0 && ni.expiresAt == 0
+}
 
-	sdb.names.Store(registrantId, name)
-	sdb.nameRegistrants.Store(nameId, registrant)
+func getRegistrantId(registrant []byte) string {
+	return hex.EncodeToString(registrant)
 }
 
-func (sdb *StateDB) deleteName(registrantId string) error {
-	name, err := sdb.trie.TryGet(append(NameRegistrantPrefix, registrantId...))
+func getNameId(name string, registrant []byte, expiresAt uint32) ([]byte, error) {
+	nameId := bytes.NewBuffer(nil)
+	nameId.Write([]byte(name))
+
+	ni := newNameInfo(registrant, expiresAt)
+	niBytes, err := ni.Bytes()
 	if err != nil {
-		return err
+		return nil, err
 	}
+	registrantHash := sha256.Sum256(niBytes)
+	nameId.Write(registrantHash[:hashPrefixLength])
+	return nameId.Bytes(), nil
+}
 
-	nameId := getNameId(string(name))
-	err = sdb.trie.TryDelete(append(NameRegistrantPrefix, registrantId...))
+func (sdb *StateDB) getRegName(name string) (string, error) {
+	iter := trie.NewIterator(sdb.trie.NodeIterator(append(NamePrefix, name...)))
+	iter.Next()
+	data := iter.Value
+	ni := new(nameInfo)
+	if len(data) > 0 {
+		buff := bytes.NewBuffer(data)
+		if err := ni.Deserialize(buff); err != nil {
+			return "", err
+		}
+	}
+
+	return ni.String(), nil
+}
+
+func (sdb *StateDB) getNameInfo(nameId []byte) (*nameInfo, error) {
+	ni := new(nameInfo)
+
+	if v, ok := sdb.names.Load(string(nameId)); ok {
+		if ni, ok := v.(*nameInfo); ok {
+			return ni, nil
+		}
+	}
+
+	enc, err := sdb.trie.TryGet(append(NamePrefix, nameId...))
+	if err != nil {
+		return nil, err
+	}
+	if len(enc) > 0 {
+		buff := bytes.NewBuffer(enc)
+		if err := ni.Deserialize(buff); err != nil {
+			return nil, err
+		}
+	}
+
+	sdb.names.Store(string(nameId), ni)
+	return ni, nil
+}
+
+func (sdb *StateDB) nameRegistration(name string, registrant []byte, expiresAt uint32) error {
+	nameId, err := getNameId(name, registrant, expiresAt)
 	if err != nil {
 		return err
 	}
 
-	err = sdb.trie.TryDelete(append(NamePrefix, nameId...))
+	ni, err := sdb.getNameInfo(nameId)
 	if err != nil {
 		return err
 	}
 
-	sdb.names.Delete(registrantId)
-	sdb.nameRegistrants.Delete(nameId)
+	if ni.Empty() {
+		ni.registrant = registrant
+		ni.expiresAt = expiresAt
+	} else {
+		return errors.New("name got registered")
+	}
+
+	return sdb.cleanupNamesAtHeight(expiresAt, nameId)
+}
+
+func (sdb *StateDB) updateNameInfo(nameId string, ni *nameInfo) error {
+	return sdb.updateName(nameId, ni)
+}
+
+func (sdb *StateDB) updateName(nameId string, ni *nameInfo) error {
+	buff := bytes.NewBuffer(nil)
+	if err := ni.Serialize(buff); err != nil {
+		return nil
+	}
```

Comment:
Duplicated code with `ni.Bytes()`. We should reuse `ni.Bytes()` which had implemented as above.

Reply:
done